examples/tutorials/additive_synthesis_tutorial.py

```diff
@@ -0,0 +1,326 @@
+SAMPLE_RATE = 16_000  # [Hz]
+
+NUM_FRAMES = int(DURATION * SAMPLE_RATE)
+
+######################################################################
+#
+
+def show(freq, amp, waveform, sample_rate, zoom=None, vol=0.1):
+    t = torch.arange(waveform.size(0)) / sample_rate
+
+    fig, axes = plt.subplots(4, 1, sharex=True)
+    axes[0].plot(t, freq)
+    axes[0].set(
+        title=f"Oscillator bank (bank size: {amp.size(-1)})",
+        ylabel="Frequency [Hz]",
+        ylim=[-0.03, None])
+    axes[1].plot(t, amp)
+    axes[1].set(
+        ylabel="Amplitude",
+        ylim=[-0.03 if torch.all(amp >= 0.0) else None, None])
+    axes[2].plot(t, waveform)
+    axes[2].set(ylabel="Waveform")
+    axes[3].specgram(waveform, Fs=sample_rate)
+    axes[3].set(
+        ylabel="Spectrogram",
+        xlabel="Time [s]",
+        xlim=[-0.01, t[-1] + 0.01])
+
+    for i in range(4):
+        axes[i].grid(True)
+    pos = axes[2].get_position()
+    plt.tight_layout()
+
+    if zoom is not None:
+        ax = fig.add_axes([pos.x0 + 0.02, pos.y0 + 0.03, pos.width / 2.5, pos.height / 2.0])
+        ax.plot(t, waveform)
+        ax.set(xlim=zoom, xticks=[], yticks=[])
+
+    waveform /= waveform.abs().max()
+    return Audio(vol * waveform, rate=sample_rate, normalize=False)
+
+######################################################################
+# Harmonics Overtones
+# -------------------
+#
+# Harmonic overtones are frequency components that are integer
+# multiple of the fundamental frequency.
+#
+# We look at how to generate the common waveforms that are used in
+# synthesizers. That is,
+#
+#  - Sawtooth wave
+#  - Square wave
+#  - Triangle wave
+#
+
+######################################################################
+# Sawtooth wave
+# ~~~~~~~~~~~~~
+#
+# `Sawtooth wave <https://en.wikipedia.org/wiki/Sawtooth_wave>`_ can be
+# expressed as the following. It contains all the integer harmonics, so
+# it is commonly sued in subtractive synthesis as well.
+#
+# .. math::
+#
+#    \begin{align*}
+#    y_t &= \sum_{k=1}^{K} A_k \sin ( 2 \pi f_k t ) \\
+#    \text{where} \\
+#    f_k &= k f_0 \\
+#    A_k &= -\frac{ (-1) ^k }{k \pi}
+#    \end{align*}
+#
+
+######################################################################
+# The following function takes fundamental frequencies and amplitudes,
+# and adds extend pitch in accordance with the formula above.
+#
+
+def sawtooth_wave(freq0, amp0, num_pitches, sample_rate):
+    freq = extend_pitch(freq0, num_pitches)
+
+    mults = [-((-1) ** i) / (PI * i) for i in range(1, 1+num_pitches)]
+    amp = extend_pitch(amp0, mults)
+    waveform = oscillator_bank(freq, amp, sample_rate=sample_rate)
+    return freq, amp, waveform
+
+
+######################################################################
+#
+# Now synthesize a waveform
+#
+
+freq0 = torch.full((NUM_FRAMES, 1), F0)
+amp0 = torch.ones((NUM_FRAMES, 1))
+freq, amp, waveform = sawtooth_wave(freq0, amp0, int(SAMPLE_RATE / F0), SAMPLE_RATE)
```

Comment:
just an fyi, a couple times in the tutorial there is a user warning `Some frequencies are above nyquist frequency. Setting the corresponding amplitude to zero. This might cause numerically unstable gradient.`

Reply:
Yeah, this happens because we are filling the entire frequency range with frequency series some of which exceed the Nyquist frequency. We could suppress the the warning, but I thought it's overkill and since we do not do any back propagation in this tutorial, so I think it's okay. We have occasional warnings (like the one from PyTorch core when creating a Tensor from read-only buffer)